File: leetcode/medium/SpiralMatrix.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        """
        Traverses an m x n matrix in spiral order.

        This solution uses four pointers to define the boundaries of the matrix:
        `top`, `bottom`, `left`, and `right`. We traverse the outer layer of the
        current boundaries and then shrink them inward until the pointers cross.

        Algorithm:
        1. Initialize boundary pointers: `top=0`, `bottom=m-1`, `left=0`, `right=n-1`.
        2. Loop as long as `top <= bottom` and `left <= right`.
        3. In each iteration of the loop, traverse one full layer:
           a. Traverse Right: Go from `left` to `right` along the `top` row. Increment `top`.
           b. Traverse Down: Go from `top` to `bottom` along the `right` column. Decrement `right`.
           c. Traverse Left: Go from `right` to `left` along the `bottom` row. Decrement `bottom`.
           d. Traverse Up: Go from `bottom` to `top` along the `left` column. Increment `left`.
        4. Add checks before the "Traverse Left" and "Traverse Up" steps to handle
           matrices that are not square (e.g., single rows or columns).

        Time Complexity: O(m * n), as we visit each element exactly once.
        Space Complexity: O(1), excluding the space for the result array.
        """
        if not matrix or not matrix[0]:
            return []

        m, n = len(matrix), len(matrix[0])
        result = []
        top, bottom = 0, m - 1
        left, right = 0, n - 1

        while top <= bottom and left <= right:
            # 1. Traverse from left to right along the top row.
            for j in range(left, right + 1):
                result.append(matrix[top][j])
            top += 1  # Move the top boundary down.

            # 2. Traverse from top to bottom along the right column.
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1  # Move the right boundary left.

            # Check if there's still a valid row to traverse.
            if top <= bottom:
                # 3. Traverse from right to left along the bottom row.
                for j in range(right, left - 1, -1):
                    result.append(matrix[bottom][j])
                bottom -= 1  # Move the bottom boundary up.

            # Check if there's still a valid column to traverse.
            if left <= right:
                # 4. Traverse from bottom to top along the left column.
                for i in range(bottom, top - 1, -1):
                    result.append(matrix[i][left])
                left += 1  # Move the left boundary right.

        return result

File: leetcode/medium/SpiralMatrix.java
import java.util.ArrayList;
import java.util.List;

/**
 * Solution class for traversing a 2D matrix in a spiral order.
 */
class Solution {
    /**
     * Given an m x n matrix, return all elements of the matrix in spiral order.
     * * @param matrix The input 2D integer array (matrix).
     * @return A List of Integers containing all elements of the matrix 
     * in spiral order.
     */
    public List<Integer> spiralOrder(int[][] matrix) {
        // Initialize an empty list to store the result
        List<Integer> result = new ArrayList<>();
        
        // Handle edge case where the matrix is null or empty
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return result;
        }

        int rows = matrix.length;
        int cols = matrix[0].length;
        
        // Define boundaries for the spiral traversal
        int top = 0;       // Top row index
        int bottom = rows - 1; // Bottom row index
        int left = 0;      // Left column index
        int right = cols - 1;  // Right column index

        // The loop continues as long as the boundaries haven't crossed
        while (top <= bottom && left <= right) {
            
            // 1. Traverse Right (from left to right along the top row)
            for (int c = left; c <= right; c++) {
                result.add(matrix[top][c]);
            }
            top++; // Move the top boundary down

            // 2. Traverse Down (from top to bottom along the rightmost column)
            if (top <= bottom) { // Check if the top boundary hasn't crossed the bottom
                for (int r = top; r <= bottom; r++) {
                    result.add(matrix[r][right]);
                }
                right--; // Move the right boundary left
            }

            // 3. Traverse Left (from right to left along the bottom row)
            if (top <= bottom && left <= right) { // Check if boundaries haven't crossed
                for (int c = right; c >= left; c--) {
                    result.add(matrix[bottom][c]);
                }
                bottom--; // Move the bottom boundary up
            }

            // 4. Traverse Up (from bottom to top along the leftmost column)
            if (top <= bottom && left <= right) { // Check if boundaries haven't crossed
                for (int r = bottom; r >= top; r--) {
                    result.add(matrix[r][left]);
                }
                left++; // Move the left boundary right
            }
        }

        return result;
    }
}